Source/Events.Processing/IEventProcessorsRegistration.cs

```diff
@@ -0,0 +1,43 @@
+// Copyright (c) Dolittle. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Threading.Tasks;
+using Dolittle.Runtime.Events.Processing.Streams;
+
+namespace Dolittle.Runtime.Events.Processing
+{
+    /// <summary>
+    /// Defines a system which controls the registration of event processors.
+    /// </summary>
+    public interface IEventProcessorsRegistration : IDisposable
+    {
+        /// <summary>
+        /// Gets a value indicating whether the registration is completed.
+        /// </summary>
+        bool Completed { get; }
+
+        /// <summary>
+        /// Gets a value indicating whether the registration succeeded or not.
+        /// </summary>
+        bool Succeeded { get; }
+
+        /// <summary>
+        /// Begins registration.
+        /// </summary>
+        /// <returns>A <see cref="Task" /> that, when resolved, returns the <see cref="EventProcessorsRegistrationResult" />.</returns>
+        Task<EventProcessorsRegistrationResult> Register();
+
+        /// <summary>
+        /// Starts the registered <see cref="StreamProcessor" />s and completes the registration.
+        /// </summary>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        Task Complete();
+
+        /// <summary>
+        /// Fail the registration.
+        /// </summary>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        Task Fail();
+    }
+}
```

Comment:
I don't fully understand this interface - and then in turn all the other implementations of it. Perhaps it helps if you explain a bit of the logic behind it? I see there is a description in the PR text, but I didn't really understand how it works.

So what is it in practice that happens in the `Register()` method? Is that the place where the implementations go their respective registry - and potentially fail if a `StreamProcessor` is already registered with that specific _id_?

It seems like the logic is supposed to be that you first call `Register()`, then either `Complete()` or `Fail()`. What happens in each of those? Do they either _start_ or _unregister_ the `StreamProcessor`s again?

In the `FilterRegistration` it seems like that is where the _definition_ is persisted to the database, which could potentially fail. So what happens if something goes wrong during the `Complete()` call?